Add information about `Flags` and `Init` modules

jwoudenberg opened the pull request:
Resolves #6 

This is a writeup of the result of a [discussion on Discourse](https://discourse.noredink.com/t/dependency-between-model-and-flags/83).

jwoudenberg committed:
Add information about `Flags` and `Init` modules

tesk9 commented on README.md:
```diff
@@ -82,21 +82,27 @@ Our Elm apps generally take this form:
 
 Inside **`View.elm`**, we define the view for our model and set up any event handlers we need.
 
+**`Flags.elm`** contains a decoder for the flags of the app. We aim to keep our decoders basic and so decode into a special `Flags` type that mirrors the structure of the flags JSON instead of the structure of the `Model` type. The `Flags` and `Model` modules should not depend on each other.
```
maybe "raw JSON"? not sure a person who isn't familiar with the flags pattern will understand "flags JSON"

tesk9 commented on README.md:
```diff
@@ -82,21 +82,27 @@ Our Elm apps generally take this form:
     - Our entry point. Contains an initial model, `Html.programWithFlags` calls and ports.
     - Compile target for `elm-make`
-    - Imports `Model`, `Update` and `View`.
+    - Imports `Init`, `Update` and `View`.
```
Won't it still need Model for the `main : Program Flags Model Msg` type signatury bit?

jwoudenberg replied:
Good point!

tesk9 reviewed (commented):
Seems 👍 . Thanks for writing up!

I expect @avh4 will have thoughts, and I wonder if would be more clear to have example code or type signatures (e.g., `init : Flags -> Model`)

jwoudenberg commented:
@tesk9 I like your idea of adding some type signatures. When I was trying this out in #15175 yesterday with @stoeffel we realised that if `Init` is importing `Flags` anyway, it might as well do the entire initialisation. That would result in a signature `String -> (Model, List (Cmd Msg))`. How do you feel about that?

**EDIT:** I pushed a change addressing your feedback and adding some proposed type definitions.

jwoudenberg committed:
Be more specific about module exports

joneshf commented on README.md:
```diff
@@ -82,34 +82,54 @@ Our Elm apps generally take this form:
 
+**`Flags.elm`** contains a decoder for the flags of the app. We aim to keep our decoders basic and so decode into a special `Flags` type that mirrors the structure of the raw JSON instead of the structure of the `Model` type. The `Flags` and `Model` modules should not depend on each other.
+
+**`Init.elm`** exports an `init` function responsible for running the decoder in `Flags` and turning the resuling `Flags` into a `Model` type. This means `init` should have type.
```
Is this last sentence necessary? The type signature is down below.

avh4 commented:
imo,  we shouldn't have `Init.elm`, and the `init` function should go in `Model`, since that's normally the only function that can create Model values.

Other that that, this seems good to me.

joneshf commented:
> That would result in a signature String -> (Model, List (Cmd Msg)). How do you feel about that?

Wouldn't that mean that `Init` now depends on `Update`? I'm not against it, just pointing it out. In fact, I think it's a good idea. It more closely matches what the "program" is.

avh4 commented:
(Here's a recent example of extracting Flags that seemed to go nicely) https://gist.github.com/avh4/ad244cdc8bc2de9085fe15238660a7aa

edit: possibly a nicer diff of the same thing: https://github.com/NoRedInk/NoRedInk/pull/15194/files#diff-5ba1bd6314b97700b2d487c2a711b85d

jwoudenberg commented:
@joneshf It would, though I hadn't realised yet because in our test case we could get away with using `msg` instead of `Msg` in `init`, so importing `Update` wasn't necessary.

@avh4 The motivation for moving `init` out of `Model` was to make it explicit that the `Model` type does not have a dependency on the `Flags` type. Because `init` clearly does have a dependency on the `Flags` module, putting it in `Model` drags that dependency in.

avh4 commented:
yeah, I think Model rightly has a dependency on Flags, because you need a Flags to create a Model.  So I think there is a clear dependency there.  And also because init is the only way to create a Model, so I think it belongs in the Model module, and putting it in a separate module means then Model has to expose a bunch of things that really could be private except that init needs them.

joneshf commented:
Do we want `init` to be the only way to create a `Model`? Wouldn't that make all tests that involve `Model` now require knowledge of the underlying JSON representation and how that relates to the `Model`?

It seems like we're missing something if we cannot make a value except by way of making some other value first and then applying a function to it.

avh4 commented:
@jwoudenberg said on discourse:
> What confuses me a bit is that a dependency of Model on Flags seems to express to me the exact opposite sentiment: It formalises a dependency of the entire app on the backend representation as expressed in Flags, which we want to avoid. Because if Flags changes then Model needs to change too, and if Model changes that affects the entire app.

If that's the concern, then what about putting `init` in `Main`?  In general, I don't think having more modules in every folder makes our code simpler.

jwoudenberg committed:
Do Model initialization in Main.elm

jwoudenberg committed:
Update dependency graph

jwoudenberg commented:
Putting initialisation in `Main.elm` also makes a lot of sense to me, and I think addresses all concerns in the discuss thread. PR has been updated to reflect this.

eeue56 commented:
Some external input, feel free to completely ignore!

> Do we want init to be the only way to create a Model? Wouldn't that make all tests that involve Model now require knowledge of the underlying JSON representation and how that relates to the Model?
> It seems like we're missing something if we cannot make a value except by way of making some other value first and then applying a function to it.

I think this is a good point and not one to miss. One nice pattern than I've found myself using more lately is to have a `defaultModel`, and then `init` simply updates the `defaultModel`. That way, my code is easily testable from the offset, without needing to write a decoder or dump some example JSON. It's better (IMO) to leave validation of JSON loading to things like schema checks, or integration tests. But unit tests which need default models are better off with a default model, rather than loading some JSON that may not be accurate with what the server gives you.

So then I would end up with:

```elm
-- Model.elm

type alias Model = {..}

defaultModel : Model
defaultModel = {..}
```

```elm
-- Flags.elm
type alias Flags = {..}
```

```elm
-- Main.elm
init : Flags -> Model
init flags =
  { defaultModel 
  | someField = flags.someValue
  }
```

eeue56 commented on README.md:
```diff
@@ -82,34 +82,47 @@ Our Elm apps generally take this form:
+- Flags.elm
+    - Contains the flags decoder
+    - Imports nothing but generalized decoders.
+    - Exports `Flags`
```
💭 this should probably be `(Flags, decodeFlags)`?

avh4 commented on images/module-dependencies.dot:
```diff
@@ -0,0 +1,8 @@
+  Model -> Update
+  Model -> View
+  Flags -> Main
+}
```
💭  I thought the png came out nicer when Main was at the top instead of at the bottom... Is there any way to force dot to do that?

ento replied:
`rank` can probably add that constraint http://graphs.grevian.org/example#example-7

joneshf replied:
I originally set a default edge attribute:

```dot
digraph {
  edge [dir=back]

  Update -> Main
  View -> Main
  ...
}
```

But, maybe that's not the most robust way...

jwoudenberg replied:
Got it to work!

avh4 commented:
✅  LGMT (didn't approve because it looks like several others may want to give opinions on the most recent change)

stoeffel commented on README.md:
```diff
@@ -82,34 +82,47 @@ Our Elm apps generally take this form:
 - Update.elm
     - Contains the `Msg` type for the view, and the update function.
     - Imports `Model`
+    - Exports `update : Msg -> Model -> (Model, List (Cmd Msg))` and `Msg`
```
I think we should clarify why we return `(Model, List (Cmd Msg))`

stoeffel commented:
I think it's not clear atm where `init` should be. I like Noah's proposal with a `defaultModel` and `init` in main for that.

jwoudenberg committed:
Add `decodeFlags` to `Flags` exports

jwoudenberg committed:
Reverse order dependency graph

jwoudenberg committed:
Expand on function of `Main.elm`

jwoudenberg commented:
I made a number of further updates to include some more of the feedback:
- The diagram has `Main` at the top again.
- The `decodeFlags` export is now listed.
- The `Main.elm` section now includes more details on how the exported values from other modules are combined, so it should be clearer it is responsible for `init` and handling the non-standard type signature for `update`.

I really like the idea of a `defaultModel` as well, especially for unit testing purposes. That seems like a separate discussion to me though, so I propose we split that off to a separate discussion, perhaps in Discourse?

stoeffel commented on README.md:
```diff
@@ -95,6 +92,16 @@ Inside **`View.elm`**, we define the view for our model and set up any event han
+
+It calls `Html.programWithFlags` with:
+- `init`, defined in `Main`, runs `Flags.decodeFlags` and turns the resulting `Flags` type into a `Model`.
+- `update` is `Update.update >> Util.batchUpdate`. That last part collapses the `List (Cmd Msg)` into a single `Cmd Msg`.
```
maybe link to https://github.com/NoRedInk/rocket-update

stoeffel commented on README.md:
```diff
@@ -95,6 +92,16 @@ Inside **`View.elm`**, we define the view for our model and set up any event han
+- `view` is simply `View.view`.
+- `subscriptions`, defined in `Main`, contains any subscriptions this app relies on.
+
+Additionally we setup ports for interop with JS in this file. We run elm-make on this file to generate a JS file that we can include elsewhere.
```
we sometimes have a `Ports.elm` though

jwoudenberg replied:
I guess that's necessary because `Update` can depend on a port and it will not be able to import them from `Main.elm`. Do we want to write down the use of a separate `Ports.elm` as the default, or something we do when needed?

jwoudenberg committed:
Add reference to rocket-update

ento reviewed (approved):
LGTM. 😻 the dependency graph. There's still the question of `Ports.elm` but that could be addressed in another PR.

jwoudenberg commented:
Thanks for all comments! Seems there are no further comments for now, so merging!
